craft_providers/multipass/multipass.py

```diff
@@ -0,0 +1,382 @@
+
+        :param source: The source path, prefixed with <name:> for a path inside
+            the instance.
+        :param destination: An IO stream to write to.
+
+        :raises MultipassError: On error.
+        """
+        command = [str(self.multipass_path), "transfer", source, "-"]
+        proc = subprocess.Popen(
+            command,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.PIPE,
+        )
+
+        # Should never happen, but pyright makes noise.
+        assert proc.stdout is not None
+
+        while True:
+            data = proc.stdout.read(chunk_size)
+            if not data:
+                break
+
+            destination.write(data)
+
+        try:
+            _, stderr = proc.communicate(timeout=30)
+        except subprocess.TimeoutExpired:
+            proc.kill()
+            _, stderr = proc.communicate()
+
+        if proc.returncode != 0:
+            raise MultipassError(
+                brief=f"Failed to transfer file {source!r}.",
+                details=errors.details_from_command_error(
+                    cmd=command, stderr=stderr, returncode=proc.returncode
+                ),
+            )
+
+    def transfer_source_io(
+        self, *, source: io.BufferedIOBase, destination: str, chunk_size: int = 4096
+    ) -> None:
+        """Transfer to destination path with source IO.
+
+        Note that this can't use std{in,out}=open(...) due to LP #1849753.
+
+        :param source: An IO stream to read from.
+        :param destination: The destination path, prefixed with <name:> for a
+            path inside the instance.
+
+        :raises MultipassError: On error.
+        """
+        command = [str(self.multipass_path), "transfer", "-", destination]
+        proc = subprocess.Popen(command, stdin=subprocess.PIPE)
+
+        # Should never happen, but pyright makes noise.
+        assert proc.stdin is not None
+
+        while True:
+            data = source.read(chunk_size)
+            if not data:
+                break
+
+            proc.stdin.write(data)
+
+        # Wait until process is complete.
+        try:
+            _, stderr = proc.communicate(timeout=30)
+        except subprocess.TimeoutExpired:
+            proc.kill()
+            _, stderr = proc.communicate()
+
+        if proc.returncode != 0:
+            raise MultipassError(
+                brief=f"Failed to transfer file to destination {destination!r}.",
+                details=errors.details_from_command_error(
+                    cmd=command, stderr=stderr, returncode=proc.returncode
+                ),
+            )
+
+    def umount(self, *, mount: str) -> None:
+        """Unmount target in VM.
+
+        :param mount: Mount point in <name>[:<path>] format, where <name> are
+            instance names, and optional <path> are mount points.  If omitted,
+            all mounts will be removed from the name instance.
```

Comment:
"from the named instance" or "from the <name> instance"?

Reply:
Fixed